BoxContentSDK/BoxContentSDK/OAuth2/BOXAuthorizationViewController.m

```diff
@@ -504,45 +507,86 @@ - (void)connection:(NSURLConnection *)connection willSendRequestForAuthenticatio
             NSError *myError = [NSError errorWithDomain:NSURLErrorDomain code:NSURLErrorUserAuthenticationRequired userInfo:nil];
-            [self failAuthenticationWithConnection:connection challenge:challenge message:message error:myError];
+            [self failAuthenticationWithURLSessionTask:task message:message error:myError];
         } else {
             self.authChallengeCycles++;
 
             //  The NTLM protocol issues a 401 as part of the negotiation process.
-            //  In iOS 8, this 401 results in NSURLAuthenticationChallenge returning a "failureResponse".
+            //  In iOS 8 and later, this 401 results in NSURLAuthenticationChallenge returning a "failureResponse".
             //  However, NSURLAuthenticationChallenge does not return this in iOS 7.
             //  As seen from the CFNetworking logs, NSURLAuthenticationChallenge handles the 401 negotiation,
             //  and the process is seamless in iOS 7.
-            //  However, in iOS 8, the last step of the NTLM protocol returns an error that
+            //  However, in iOS 8 and later, the last step of the NTLM protocol returns an error that
             //  the developer must handle (as seen from CFNetworking and our app logs).
+            //  We only support versions of iOS newer than iOS 8.
             BOOL shouldApplyiOS8_NTLM_WAR = (self.isNTLMAuth &&
                                              [challenge previousFailureCount] > 0 &&
-                                             SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"8.0") &&
                                              (self.ntlmAuthFailures < kMaxNTLMAuthFailuresPriorToExit));
 
             // Handle the authentication challenge as a basic HTTP authentication challenge
             // (certificate-based, among other methods, is not currently supported).
             if (shouldApplyiOS8_NTLM_WAR) {
                 BOXLog(@"Applying iOS8 NTLM WAR, ntlmAuthFaulures = %ld", (long)self.ntlmAuthFailures);
                 self.ntlmAuthFailures++;
-                [[challenge sender] continueWithoutCredentialForAuthenticationChallenge:challenge];
+                if (completionHandler) {
+                    completionHandler(NSURLSessionAuthChallengeUseCredential, NULL);
+                } else {
+                    [[challenge sender] continueWithoutCredentialForAuthenticationChallenge:challenge];
+                }
             } else if (([challenge previousFailureCount] > 0 &&
                         self.authenticationChallengeCredential == nil) ||
                        [challenge previousFailureCount] > 1) {
                 BOXLog(@"Have %ld previous failures", (long)[challenge previousFailureCount]);
+                if (completionHandler) {
+                    completionHandler(NSURLSessionAuthChallengeCancelAuthenticationChallenge, NULL);
```

Comment:
not sure what's the impact of calling completionHandler earlier in the flow, should we call it at the very end, i.e. after calling failAuthenticationWithURLSessionTask?